internal/liquids/nova/usage.go

```diff
@@ -0,0 +1,203 @@
+		"server_group_members": {
+			Quota: &absoluteLimits.MaxServerGroupMembers,
+			PerAZ: map[liquid.AvailabilityZone]*liquid.AZResourceUsageReport{liquid.AvailabilityZoneAny: {Usage: totalServerGroupMembersUsed}},
+		},
+	}
+	for flavorName, flavorLimits := range limitsData.Limits.AbsolutePerFlavor {
+		if l.IgnoreFlavor(flavorName) {
+			continue
+		}
+		resourceName := ResourceNameForFlavor(flavorName)
+		resources[resourceName] = &liquid.ResourceUsageReport{
+			Quota: &flavorLimits.MaxTotalInstances,
+			PerAZ: liquid.AZResourceUsageReport{Usage: flavorLimits.TotalInstancesUsed}.PrepareForBreakdownInto(req.AllAZs),
+		}
+	}
+	for hwVersion, limits := range limitsData.Limits.AbsolutePerHWVersion {
+		if l.hasPooledResource.Get()[hwVersion]["cores"] {
+			resources[l.pooledResourceName(hwVersion, "cores")] = &liquid.ResourceUsageReport{
+				Quota: &limits.MaxTotalCores,
+				PerAZ: liquid.AZResourceUsageReport{Usage: limits.TotalCoresUsed}.PrepareForBreakdownInto(req.AllAZs),
+			}
+		}
+		if l.hasPooledResource.Get()[hwVersion]["instances"] {
+			resources[l.pooledResourceName(hwVersion, "instances")] = &liquid.ResourceUsageReport{
+				Quota: &limits.MaxTotalInstances,
+				PerAZ: liquid.AZResourceUsageReport{Usage: limits.TotalInstancesUsed}.PrepareForBreakdownInto(req.AllAZs),
+			}
+		}
+		if l.hasPooledResource.Get()[hwVersion]["ram"] {
+			resources[l.pooledResourceName(hwVersion, "ram")] = &liquid.ResourceUsageReport{
+				Quota: &limits.MaxTotalRAMSize,
+				PerAZ: liquid.AZResourceUsageReport{Usage: limits.TotalRAMUsed}.PrepareForBreakdownInto(req.AllAZs),
+			}
+		}
+	}
+
+	// Nova does not have a native API for AZ-aware usage reporting,
+	// so we will obtain AZ-aware usage stats by counting up all subresources,
+	// even if we don't end up showing them in the API
+	allSubresources, err := l.buildInstanceSubresources(ctx, projectUUID)
+	if err != nil {
+		return liquid.ServiceUsageReport{}, fmt.Errorf("while collecting instance data: %w", err)
+	}
+
+	for _, subres := range allSubresources {
+		var attrs SubresourceAttributes
+		if err = json.Unmarshal(subres.Attributes, &attrs); err != nil { // TODO: Unmarshalling here + marshalling when building the subresources seems inefficient
+			return liquid.ServiceUsageReport{}, err
+		}
+
+		az := attrs.AZ
+
+		if l.IgnoreFlavor(attrs.FlavorName) {
+			continue
+		}
+
+		// use separate instance resource if we have a matching "instances_$FLAVOR" resource
+		instanceResourceName := ResourceNameForFlavor(attrs.FlavorName)
+		isPooled := false
+		if _, exists := resources[instanceResourceName]; !exists {
+			// otherwise used the appropriate pooled instance resource
+			isPooled = true
+			instanceResourceName = l.pooledResourceName(attrs.HWVersion, "instances")
+		}
+
+		// count subresource towards "instances" (or separate instance resource)
+		resources[instanceResourceName].AddLocalizedUsage(az, 1)
+		if l.WithSubresources {
+			azData := UsageInAZ(resources[instanceResourceName], az)
```

Comment:
It demonstrates good instinct that you're being defensive about whether the AZResourceReport exists or not. But in this case, `AddLocalizedUsage()` already guarantees it, so this can be greatly simplified to:
```suggestion
			azData := resources[instanceResourceName].PerAZ[az] // and drop the UsageInAZ function
```